crates/dbsp/src/operator/neighborhood.rs

```diff
@@ -0,0 +1,696 @@
+/// ```
+pub type Neighborhood<K, V, R> = OrdIndexedZSet<isize, (K, V), R>;
+
+/// Neighborhood descriptor represents a request from the user to
+/// output a specific neighborhood.  The neighborhood is defined in
+/// terms of its central point (`anchor`) and the number of rows
+/// preceding and following the anchor to output.
+#[derive(Clone)]
+pub struct NeighborhoodDescr<K, V> {
+    anchor: (K, V),
+    before: usize,
+    after: usize,
+}
+
+/// Stream of neighborhoods output by the [`Stream::neighborhood`] operator.
+pub type NeighborhoodStream<K, V, R> = Stream<RootCircuit, Neighborhood<K, V, R>>;
+
+/// Stream of neighborhood descriptors from
+pub type NeighborhoodDescrStream<K, V> = Stream<RootCircuit, Option<NeighborhoodDescr<K, V>>>;
+
+impl<K, V> NeighborhoodDescr<K, V> {
+    pub fn new(k: K, v: V, before: usize, after: usize) -> Self {
+        Self {
+            anchor: (k, v),
+            before,
+            after,
+        }
+    }
+}
+
+impl<B> Stream<RootCircuit, B>
+where
+    B: IndexedZSet + Send,
+    B::R: ZRingValue,
+{
+    /// Returns a small contiguous range of rows ([`Neighborhood`]) of the input
+    /// table.
+    ///
+    /// This operator helps to visualize the contents of the input table in a
+    /// UI.  The UI client may not have enough throughput/memory to store the entire
+    /// table, and will instead limit its state to a small range of rows that fit
+    /// in the screen.  We specify such a range, or _neighborhood_, in terms of its
+    /// center (or "anchor"), and the number of rows preceding and following the
+    /// anchor (see [`NeighborhoodDescr`]).  The user may be interested in a static
+    /// snapshot of the neighborhood or in a changing view.  Both modes are
+    /// supported by this operator (see the `reset` argument).  The output of the
+    /// operator is a stream of [`Neighborhood`]s.
+    ///
+    /// # Arguments
+    ///
+    /// * `self` - a stream of changes to an indexed Z-set.
+    ///
+    /// * `neighborhood_descr` - contains the neighborhood descriptor to
+    ///   evaluate at every clock tick.  Set to `None` to disable the
+    ///   operator (it will output an empty neighborhood).
+    ///
+    /// * `reset` - controls whether the operator outputs the the entire
+    ///   neighborhood or an incrmental change to the neighborhood.  Set to
+    ///   `true` to request the complete neighborhood to be output at the end
+    ///   of the current clock cycle or `false` to output the incremental
+    ///   change to the neighborhood relative to the previous clock cycle.
+    ///
+    /// # Output
+    ///
+    /// The output neighborhood will contain rows with indexes between
+    /// `-descr.before` and `descr.after - 1`.  Row 0 is the anchor row, i.e.,
+    /// is the first row in the input stream greater than or equal to
+    /// `descr.anchor`.  If there is no such row in the input stream, then
+    /// the neighborhood will only contain negative indexes.
```

Comment:
I don't see why there should be no positive indexes.

Reply:
Maybe the sentence is not clear.  If all rows in the table are smaller than the anchor, then the neighborhood will only contain negative indexes.  I'll rephrase.